net_sink.go

```diff
@@ -312,34 +319,51 @@ func (s *netSink) run() {
 			// Drop through in case retryc has nothing.
 		}
 
+		// send batched outc data anytime we're full and can't batch anymore
+		if len(batch) == cap(batch) {
+			batch = s.sendBatch(batch)
+		}
+
+		// flush to outc and batch and/or send outc
 		select {
-		case <-t.C:
-			s.flush()
 		case done := <-s.doFlush:
-			// Only flush pending buffers, this prevents an issue where
-			// continuous writes prevent the flush loop from exiting.
-			//
-			// If there is an error writeToConn() will set the conn to
-			// nil thus breaking the loop.
-			//
-			n := len(s.outc)
-			for i := 0; i < n && s.conn != nil; i++ {
-				buf := <-s.outc
-				if err := s.writeToConn(buf); err != nil {
-					s.retryc <- buf
-					continue
-				}
-				putBuffer(buf)
-			}
+			// send batched outc data
+			batch = s.sendBatch(batch)
 			close(done)
 		case buf := <-s.outc:
-			if err := s.writeToConn(buf); err != nil {
-				s.retryc <- buf
+			if isBatchEnabled {
+				// batch outc data and rely on the doFlush or max capacity to send
+				batch = append(batch, *buf)
 				continue
 			}
-			putBuffer(buf)
+			// send outc data immediately
+			if err := s.send(buf); err == nil {
+				putBuffer(buf)
+			}
+		case <-t.C:
+			s.flush() // from a higher level, stats are written to s.bufWriter.buf at GOSTATS_FLUSH_INTERVAL_SECONDS (or adhoc for Timers). this flushes them to outc every t.C tick
+		}
+	}
+}
+
+func (s *netSink) send(buf *bytes.Buffer) error {
+	if err := s.writeToConn(buf); err != nil {
+		s.retryc <- buf
+		return err
+	}
+	return nil
+}
+
+func (s *netSink) sendBatch(batch []bytes.Buffer) []bytes.Buffer {
+	n := len(batch)
+	for i := 0; i < n && s.conn != nil; i++ {
+		buf := batch[i]
+		// todo can we just send all the batches in one call?
+		if err := s.send(&buf); err == nil {
+			putBuffer(&buf)
 		}
 	}
+	return batch[:0]
```

Comment:
If there is an error in `send` this will cause any following batched metrics to be dropped. Might be better to keep them queued. Additionally, this prevents us from putting any buffers received after the error back into the buffer pool (which is only a problem if we decide to keep the existing logic).

Reply:
If there's an error in send, the metrics will be written to the retryc channel so nothing should be dropped i think, despite clearing it from the batch. In the current state, that retry handling will escape batching altogether, will think on this.